### PR TITLE
schedule: make region scheduler and checkers aware of placement rules

### PR DESCRIPTION
**Pull request by disksing**
### What problem does this PR solve? 
Continue merge placement rules.

### What is changed and how it works?
- skip strict label match if placement rules enabled
- disallow merge regions that across different rules
- replace learnerChecker and replicaChecker with ruleChecker if placement rules enabled
- replace distinctScoreFilter with RuleFitFilter if placement rules enabled
- add tests

### Check List 

Tests 
 - Unit test

### CODE REVIEW

File: server/cluster/cluster.go [hunk 1]
```diff
@@ -200,6 +200,13 @@ func (c *RaftCluster) Start(s Server) error {
 		return nil
 	}
 
+	defaultReplicas := c.opt.GetMaxReplicas()
+	defaultLocationLabels := c.opt.GetLocationLabels()
+	c.ruleManager, err = placement.NewRuleManager(c.storage, defaultReplicas, defaultLocationLabels)
```
Comment (rleungx) [inline server/cluster/cluster.go]: When we disable the placement rules, will it affect the start process?

File: server/schedulers/balance_test.go [hunk 2]
```diff
@@ -737,6 +749,13 @@ func (s *testBalanceRegionSchedulerSuite) TestReplacePendingRegion(c *C) {
 	sb, err := schedule.CreateScheduler(BalanceRegionType, oc, core.NewStorage(kv.NewMemoryKV()), schedule.ConfigSliceDecoder(BalanceRegionType, []string{"", ""}))
 	c.Assert(err, IsNil)
 
+	s.checkReplacePendingRegion(c, tc, opt, sb)
+	opt.EnablePlacementRules = true
+	s.checkReplacePendingRegion(c, tc, opt, sb)
+}
+
+func (s *testBalanceRegionSchedulerSuite) checkReplacePendingRegion(c *C, tc *mockcluster.Cluster, opt *mockoption.ScheduleOptions, sb schedule.Scheduler) {
+
```
Comment (rleungx) [inline server/schedulers/balance_test.go]: empty line

File: server/cluster/cluster.go (hunk 1 above)
Comment (disksing) [reply server/cluster/cluster.go]: It is always loaded no matter enabled or not

File: server/schedule/checker_controller.go [hunk 3]
```diff
@@ -49,15 +52,25 @@ func (c *CheckerController) CheckRegion(region *core.RegionInfo) (bool, []*opera
...
+			}
+		}
+	} else {
+		if opController.OperatorCount(operator.OpReplica) < c.cluster.GetReplicaScheduleLimit() {
+			checkerIsBusy = false
+			if op := c.ruleChecker.Check(region); op != nil {
+				return checkerIsBusy, []*operator.Operator{op}
+			}
 		}
 	}
```
[the remark this answers is not in the document]
Comment (disksing) [reply server/schedule/checker_controller.go]: good idea

**Comment by disksing** [comment]
Please take a look @rleungx

File: server/cluster/cluster.go (hunk 1 above)
Comment (disksing) [reply server/cluster/cluster.go]: @rleungx fixed.

**Pull request merged**